core/src/main/scala/pureconfig/ConfigSource.scala

```diff
@@ -0,0 +1,320 @@
+package pureconfig
+
+import java.io.File
+import java.net.URL
+import java.nio.file.Path
+
+import scala.reflect.ClassTag
+
+import com.typesafe.config._
+import pureconfig.ConfigReader.Result
+import pureconfig.backend.{ ConfigFactoryWrapper, PathUtil }
+import pureconfig.error.{ CannotRead, ConfigReaderException, ConfigReaderFailures }
+
+/**
+ * A representation of a source from which `ConfigValue`s can be loaded, such as a file or a URL.
+ *
+ * A source allows users to load configs from this source as any type for which a `ConfigReader` is
+ * available. Raw configs can also be retrieved as a `ConfigValue`, a `ConfigCursor` or a
+ * `FluentConfigCursor`. Before using any of the loading methods described, Users can opt to focus
+ * on a specific part of a config by specifying a namespace.
+ *
+ * All config loading methods are lazy and defer resolution of references until needed.
+ */
+trait ConfigSource {
```

Comment:
(Here's another aesthetics comment. Feel free to ignore.)

Reading code which uses `ConfigSource` makes me wish the name of this `trait` didn't include `Config`; I wish it were just `Source`.  `I don't think the `Config` bit helps clarify or disambiguate in a serious way. I think it's unlikely that `scala.io.Source` is going to be used in the same context frequently.

```
    ConfigSource.applicationConf(
      files.map(ConfigSource.file(_).optional)
        .foldLeft(ConfigSource.empty)(_.withFallback(_)))
      .loadOrThrow[Config]
```

feels better to me as

```
    Source.applicationConf(
      files.map(Source.file(_).optional)
        .foldLeft(Source.empty)(_.withFallback(_)))
      .loadOrThrow[Config]
```

Reply:
I don't like this for the same reason as naming `load` above. While we can make a case for this not being confused with `scala.io.Source`, if every library decided to add their own `Source` class things would get very messy. I also don't think it differs much from all the other `Config*` classes (cursor, reader, writer).

Reply:
I also prefer `ConfigSource` for consistency with the other types in PureConfig, and because there are other `Source`s in Scala libraries in the wild (there's at least `scala.io.Source`, `Source` in Akka Streams and `Source` in Scalameta)